server/src/main/java/io/crate/types/DataType.java

```diff
@@ -247,7 +247,7 @@ public final StorageSupport<? super T> storageSupportSafe() {
 
 
     @Nullable
-    public final ValueIndexer<? super T> valueIndexer(RelationName table,
+    public final ValueIndexer<? super T> valueIndexer(DocTableInfo table,
```

Comment:
ah ok, I recalled why I didn't want to do "direct injection" - in case of nested indexers, need to propagate it. 

So it will be another one method in the ValueIndexer, empty by default, non-empty for Object/Array/Dyamic indexers, and propagating writer to all nested ObjectIndexers

Reply:
I added soureKeyWriter to the indexValue and  then it gets propagated to all nested (or newly created by DynamicIndexer) ObjectIndexers. Thanks for the suggestion!

I rebased the PR